### PR TITLE
Add Segmented control component

### PR DESCRIPTION
**Pull request by abhigyanghosh30**
## Done

- Porting Segmented Controls into React Components

### Storybook

To see rendered examples of all react-components, run:

```shell
yarn start
```

### QA in your project

from `react-components` run:

```shell
yarn build
npm pack
```

Install the resulting tarball in your project with:

```shell
yarn add <path-to-tarball>
```

### QA steps

- Visit https://react-components-896.demos.haus/?path=/story/segmented-control--default-story
- Verify everything works

## Fixes

Fixes: # .

### CODE REVIEW

**Comment by bartaz** [comment]
Thanks, that's a good start. But I think we need to make the component more minimal.

The Segmented Control is just the button, not the content it controls:

image

I know that the examples in Vanilla present it with tab content below, but it's just to demonstrate the functionality. In practice segmented control may switch something completely different, in different part of the UI.

So this component visually should be just a set of buttons, but it needs some onChange callback, so it can trigger an external function when user changes something.

```
<SegmentedControl
   buttons={[{label: "OLM", value: "olm", icon: "information"}, {label: "Test", value: "test", icon: "warning"}] }
   onChange={(value) => console.log("User selected ${value}")}
   dense={true}
/>
```

I don't know if `hasIcon` is needed? Maybe if props define icon it would be shown? Does it need a separate prop?

File: src/components/SegmentedControl/SegmentedControl.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import classNames from "classnames";
+import Icon, { ICONS } from "../Icon";
+import React, { HTMLProps, ReactNode, useRef, useState } from "react";
```
Comment (bartaz) [inline src/components/SegmentedControl/SegmentedControl.tsx:3]: Build/lint seems to fail in this. 
```suggestion
import React, { HTMLProps, useRef, useState } from "react";
```

**Review by bartaz — CHANGES_REQUESTED** [review]
@abhigyanghosh30 Would you have time to address those, or should we just continue ourselves?

Let us know if you can continue working on this. It's fine if you don't!

File: src/components/SegmentedControl/SegmentedControl.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  dense?: boolean;
+};
+
+const SegmentedControl = <P,>({
+  className,
+  tabs,
+  dense = false,
+}: Props<P>): JSX.Element => {
+  const [currentTab, changeTab] = useState(0);
```
[the remark this answers is not in the document]
Comment (abhigyanghosh30) [reply src/components/SegmentedControl/SegmentedControl.tsx:42]: I'll make the changes and get back to you

[the remark this answers is not in the document]
Comment (abhigyanghosh30) [reply src/components/SegmentedControl/SegmentedControl.tsx:42]: > There is no tabs now, so this doesn't seem to be needed anymore.
> 
> We should be able to pass some `onChange` event handler to Segmented Control component, so whenever a button is clicked the function passed to `onChange` would be called with correct value.

This function is there to shift the focused button in the group on inputs such as arrow keys. It is an accessibility requirement I think

**Comment by huwshimi** [comment]
There's also an implementation of the segmented control in MAAS that could be used as a reference: https://github.com/canonical/maas-ui/blob/main/src/app/base/components/SegmentedControl/SegmentedControl.tsx, though that one doesn't support custom button/link elements (in JAAS we change the URL query string with a segmented control so we need to use anchors).

**Comment by bartaz** [comment]
> There's also an implementation of the segmented control in MAAS that could be used as a reference: https://github.com/canonical/maas-ui/blob/main/src/app/base/components/SegmentedControl/SegmentedControl.tsx, though that one doesn't support custom button/link elements (in JAAS we change the URL query string with a segmented control so we need to use anchors).

Sounds like it would be more flexible to implement it in "material ui" style like:

```jsx
<SegmentedControl>
   <Button>Something</Button>
   <Link className="p-button">Something else</Link>
</SegmentedControl>
```

So the component itself won't need direct support for all the various options apps may need.

On the other hand it would make it more difficult to add proper styling class names and functionality of selecting buttons if children are not directly handled by parent…

**Comment by huwshimi** [comment]
> > There's also an implementation of the segmented control in MAAS that could be used as a reference: https://github.com/canonical/maas-ui/blob/main/src/app/base/components/SegmentedControl/SegmentedControl.tsx, though that one doesn't support custom button/link elements (in JAAS we change the URL query string with a segmented control so we need to use anchors).
> 
> Sounds like it would be more flexible to implement it in "material ui" style like:
> 
> ```js
> <SegmentedControl>
>    <Button>Something</Button>
>    <Link className="p-button">Something else</Link>
> </SegmentedControl>
> ```
> 
> So the component itself won't need direct support for all the various options apps may need.
> 
> On the other hand it would make it more difficult to add proper styling class names and functionality of selecting buttons if children are not directly handled by parent…

I think can just follow the pattern we use elsewhere where the component can take a prop which sets the component/element type and then a generic for the additional props. This is a good example: https://github.com/canonical/react-components/blob/main/src/components/Tabs/Tabs.tsx#L38

**Comment by bartaz** [comment]
Unfortunately I think we need to close this one.
There is currently no clear vision on how this component should be done both from design system point of view (there have been discussions of "segmented control" being different component from "button group") and from React implementation point of view (component with props vs composable components).

We can revisit it later when we have more clarity.

**Pull request closed without merging**